### PR TITLE
C# source generator for docopt usage

### PR DESCRIPTION
**Pull request by atifaziz**
This PR implements a [C# source generator](https://devblogs.microsoft.com/dotnet/introducing-c-source-generators/) for docopt. It generates strong-typed C# given a `*.docopt.txt` file.

There is an example of **NavalFate** using the source generator.

### CODE REVIEW

**Comment by atifaziz** [comment]
There is no concept or specification for the final shape of this feature. In its first draft, the source generator simply generates C# source code similar to the T4 template, but it would be a shame to leave it there. It can and should go much further; it can be a lot smarter given that the source generator has access to the syntactic and semantic model of a project at generation-time.

The PR should to help to explore and develop the feature.

**Comment by voieducode** [comment]
I love this - lots to unpack for me, I will take it for a spin!

**Comment by atifaziz** [comment]
> I love this - lots to unpack for me, I will take it for a spin!

Cool and needless to say, feature ideas welcome!

I am not sure how well this is doable/mappable, but I'd like to toy with the idea that the source generate can generate code to bind the following usage:

    naval_fate ship new <name>...
    naval_fate ship <name> move <x> <y> [--speed=<kn>]
    naval_fate ship shoot <x> <y>
    naval_fate mine (set|remove) <x> <y> [--moored|--drifting]

to the following method signatures:

```c#
int ShipNew(string[] name);
int ShipMove(string name, int x, int y, int speed);
int ShipShoot(int x, int y, int);
int Mine(Choice<Set, Remove> action, int x, int y, Choice<Set, Remove>? disposition);

// generated types:

enum Set { Set };
enum Remove { Remove };
enum Moored { Moored };
enum Drifting { Drifting }
```

**Comment by atifaziz** [comment]
With 41bf0c0c1f55fca1def81a4305b14d865e18646b and 54825a7f8070f67d0df90313e8293cd01a59f650, the source generator now creates a static version of the pattern tree created from the usage text:

```c#
// Required:
//   Either:
//     Required:
//       Command(ship, False) -> CommandNode ship Bool
//       Command(new, False) -> CommandNode new Bool
//       OneOrMore:
//         Argument(<name>, []) -> ArgumentNode <name> List
//     Required:
//       Command(ship, False) -> CommandNode ship Bool
//       Argument(<name>, []) -> ArgumentNode <name> List
//       Command(move, False) -> CommandNode move Bool
//       Argument(<x>, ) -> ArgumentNode <x> String
//       Argument(<y>, ) -> ArgumentNode <y> String
//       Optional:
//         Option(,--speed,1,10) -> OptionNode speed String
//     Required:
//       Command(ship, False) -> CommandNode ship Bool
//       Command(shoot, False) -> CommandNode shoot Bool
//       Argument(<x>, ) -> ArgumentNode <x> String
//       Argument(<y>, ) -> ArgumentNode <y> String
//     Required:
//       Command(mine, False) -> CommandNode mine Bool
//       Required:
//         Either:
//           Command(set, False) -> CommandNode set Bool
//           Command(remove, False) -> CommandNode remove Bool
//       Argument(<x>, ) -> ArgumentNode <x> String
//       Argument(<y>, ) -> ArgumentNode <y> String
//       Optional:
//         Either:
//           Option(,--moored,0,False) -> OptionNode moored Bool
//           Option(,--drifting,0,False) -> OptionNode drifting Bool
//     Required:
//       Required:
//         Option(-h,--help,0,False) -> OptionNode help Bool
//     Required:
//       Option(,--version,0,False) -> OptionNode version Bool

static readonly Pattern Pattern =
    new Required(ImmutableArray.Create<Pattern>(
        new Either(ImmutableArray.Create<Pattern>(
            new Required(ImmutableArray.Create<Pattern>(
                new Command("ship"),
                new Command("new"),
                new OneOrMore(
                    new Argument("<name>", null)))),
            new Required(ImmutableArray.Create<Pattern>(
                new Command("ship"),
                new Argument("<name>", null),
                new Command("move"),
                new Argument("<x>", null),
                new Argument("<y>", null),
                new Optional(ImmutableArray.Create<Pattern>(
                    new Option("", "--speed", 1, null))))),
            new Required(ImmutableArray.Create<Pattern>(
                new Command("ship"),
                new Command("shoot"),
                new Argument("<x>", null),
                new Argument("<y>", null))),
            new Required(ImmutableArray.Create<Pattern>(
                new Command("mine"),
                new Required(ImmutableArray.Create<Pattern>(
                    new Either(ImmutableArray.Create<Pattern>(
                        new Command("set"),
                        new Command("remove"))))),
                new Argument("<x>", null),
                new Argument("<y>", null),
                new Optional(ImmutableArray.Create<Pattern>(
                    new Either(ImmutableArray.Create<Pattern>(
                        new Option("", "--moored", 0, null),
                        new Option("", "--drifting", 0, null))))))),
            new Required(ImmutableArray.Create<Pattern>(
                new Required(ImmutableArray.Create<Pattern>(
                    new Option("-h", "--help", 0, null))))),
            new Required(ImmutableArray.Create<Pattern>(
                new Option("", "--version", 0, null)))))));
```

This means that there is _zero_ run-time cost of parsing the document usage, which is now all passed to build-time. With the above tree, one technically only needs `ParseArgv`:

https://github.com/docopt/docopt.net/blob/ed57edb35a23a89282375500f8dfa5da84439d07/src/DocoptNet/Docopt.cs#L166-L167

and the `PatternMatcher` added with PR #82 for the final bit. The runtime cost of these two functions should be very minimal. Ideally, even the `PatternMatcher` can be statically computed/generated code from the tree, one downside of which would be maintenance of duplicate code (one used for **DocoptNet** as library code and the other for source generation). The duplication could be removed by some intermediate language that interpreted at run-time but can also be using to generate code, but that might be too fancy for patterns that are not going to change much given that the docopt specification has been stable for a very long time now.

**Comment by atifaziz** [comment]
0100f0f327f16eaeeb30336a7c9addc7f3281d56 adds the first generated version of `Docopt.Apply` with the entire pattern matching tree in-lined as code. During development, I will be sharing [the generated code](https://github.com/docopt/docopt.net/tree/7f029252ade3616eeb702faeef9786ed8fc0ea38/eg/SourceGenerator/NavalFate/obj/Generated/DocoptNet.CodeGeneration/DocoptNet.CodeGeneration.SourceGenerator) for those interested in following the progress or providing feedback.

At this stage, I am not even sure the generated code works so the next step is to add tests, fix bug and then refactor (with tests avoiding regressions).

**Comment by voieducode** [comment]
We are not there yet, but I wonder how easily will this approach map to F#.

**Comment by atifaziz** [comment]
> We are not there yet, but I wonder how easily will this approach map to F#.

I don't believe the mapping will be straightforward at all since F# type providers (assuming that's what you had in mind) and C# source generators work at completely different levels. Implementation-wise, there would be practically no overlap.

For some interesting discussion and comparisons, see fsharp/fslang-suggestions#864.

**Comment by atifaziz** [comment]
The code generated by the source generator now passes all 176 of [language agnostics tests](https://github.com/docopt/docopt.net/blob/42139b7af1bad74f82f6eca75e19ac028bc65ac6/tests/LanguageAgnosticTests/testcases.docopt)! What's more, I've ported a number of the [original docopt reference examples](https://github.com/docopt/docopt/tree/20b9c4ffec71d17cee9fd963238c8ec240905b65/examples) with source generation and they behave identically ([with some adjustments](https://github.com/docopt/docopt.net/pull/77/commits/cb93a0ebc92fdc2d2136cc1864c024ce0dbe75c2)). This is largely due to the source generator and the runtime version sharing the same code base for pattern matching and processing of arguments.

The source generation tests run a full integration test that produce the code, compile to memory, load the assembly and test parsed arguments. This takes some time (between 30 to 50 seconds versus <1 for all other tests) and probably need to be classified as integration rather than unit tests.

Many polishing touches aside, I think we are now in a position to discuss what an MVP looks like for this feature/PR to come out of draft.

**Comment by atifaziz** [comment]
This PR is now published.

The latest changes have dramatically improved and reduced the size of the generated code. It's even quite readable and defers to as much static code as possible. There are still many polishing touches that can be made and `FIXME` comments that can be addressed, but I feel that the feature is an MVP in its current state and the rest of improvements can be made post-merge.

The code generator now produces a class with an `Arguments` suffix to the file name. For example, in the case of NavalFate, the usage document is in the plain text file called `Program.docopt.txt`. This will generate a class called `ProgramArguments`. It will be in the same namespace as the project's root namespace (more customisation of this can be done in the future).

The class will contain two string constants: `Help` and `Usage`. The former contains the content the of the file and the latter just the usage section, which is typically printed to the console/terminal in the event of an error.

The arguments class will contain properties for each command, positional argument and option:

```c#
partial class ProgramArguments
{
    public bool CmdShip { get; private set; }
    public bool CmdNew { get; private set; }
    public StringList ArgName { get; private set; } = StringList.Empty;
    public bool CmdMove { get; private set; }
    public string? ArgX { get; private set; }
    public string? ArgY { get; private set; }
    public string OptSpeed { get; private set; } = "10";
    public bool CmdShoot { get; private set; }
    public bool CmdMine { get; private set; }
    public bool CmdSet { get; private set; }
    public bool CmdRemove { get; private set; }
    public bool OptMoored { get; private set; }
    public bool OptDrifting { get; private set; }
    public bool OptHelp { get; private set; }
    public bool OptVersion { get; private set; }
}
```

Notice that all properties are as strong-typed as possible; that is, there's no use of `ValueObject` or `Value`. The properties are also directly initialised with defaults where applicable (such as in the case of `--speed` or `OptSpeed`). In the actual generated code, the above properties also have a [summary doc-comment element](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/xmldoc/recommended-tags#summary) (omitted above for brevity), which can help to link to the original command/argument/option and also show up as a tooltip in IDEs.

A single public and static method called `Apply` is generated that takes a sequence of strings and returns an instance of the arguments class with the respective properties populated:

```c#
public static ProgramArguments Apply(IEnumerable<string> args,
                                     bool help = true, object? version = null,
                                     bool optionsFirst = false, bool exit = false)
```

The generated class can also be treated/viewed as a general sequence of commands, arguments and options since it implements `IEnumerable<KeyValuePair<string, object?>>`:

```c#
partial class ProgramArguments : IEnumerable<KeyValuePair<string, object?>>
{
    // ...
}

This provides static and generic semantics should the latter be still useful, such as for logging,

I encourage you to check out all the examples added. I have also ported [the git examples](https://github.com/docopt/docopt/tree/20b9c4ffec71d17cee9fd963238c8ec240905b65/examples/git) from the reference implementation. Two examples are disabled until a known issue with potential duplicates can be addressed. For example, options that vary just in casing like `-d` and `-D` map to the same property `OptD` at the moment.

**Comment by voieducode** [comment]
The implementation is deeper and the output cleaner. Exciting stuff, give me a bit of time to review this.

**Comment by atifaziz** [comment]
> give me a bit of time to review this.

Absolutely and please feel free to challenge. I'll also be happy to expand on any changes that may not be so obvious and try to drop some more notes that could help with the review. There are lots of changes files and while that may seem overwhelming, many belong to examples and generated code.

**Comment by atifaziz** [comment]
### Notes for reviewers

There are a large number (75+) of files changed/added in this PR, but keep in mind that 50 of those belong to [examples added (from the reference Python implementation)](https://github.com/docopt/docopt/tree/20b9c4ffec71d17cee9fd963238c8ec240905b65/examples) that helped to shape & drive the generated code and testing:

    PS> git diff --name-only "$(git merge-base --fork-point master)..HEAD" eg/ | grep -n .

The examples include the generated code and technically don't need to be committed/tracked. They were added nonetheless to help with online review as well as regressions during development. They should/will be removed before the final merge of this PR. There are 14 such files:

    PS> git diff --name-only "$(git merge-base --fork-point master)..HEAD" eg/*/obj/* | grep -n .

When you exclude the examples, then changes to review only amount to a third (~25 files):

    PS> git diff --name-only "$(git merge-base --fork-point master)..HEAD" | grep -vF eg/SourceGenerator/ | grep -n .

There are even fewer when you consider core project changes (i.e. excluding tests):

    PS> git diff --name-only "$(git merge-base --fork-point master)..HEAD" | grep -F src/ | grep -n .
    1:src/DocoptNet.CodeGeneration/AnalyzerReleases.Shipped.md
    2:src/DocoptNet.CodeGeneration/AnalyzerReleases.Unshipped.md
    3:src/DocoptNet.CodeGeneration/CSharpSourceBuilder.cs
    4:src/DocoptNet.CodeGeneration/DocoptNet.CodeGeneration.csproj
    5:src/DocoptNet.CodeGeneration/DocoptNet.CodeGeneration.targets
    6:src/DocoptNet.CodeGeneration/Extensions.cs
    7:src/DocoptNet.CodeGeneration/SourceGenerator.cs
    8:src/DocoptNet.CodeGeneration/StringBuilderSourceText.cs
    9:src/DocoptNet/BranchPatterns.cs
    10:src/DocoptNet/Docopt.cs
    11:src/DocoptNet/GenerateCodeHelper.cs
    12:src/DocoptNet/GeneratedSourceModule.cs
    13:src/DocoptNet/Pattern.cs
    14:src/DocoptNet/PatternMatcher.cs
    15:src/DocoptNet/ValueObject.cs

In the very core project (i.e. excluding the source generator project), the bulk of the changes went into `PatternMatcher.cs`:

    PS> git diff --numstat "$(git merge-base --fork-point master)..HEAD" | grep -F src/DocoptNet/
    6       1       src/DocoptNet/BranchPatterns.cs
    1       1       src/DocoptNet/Docopt.cs
    27      5       src/DocoptNet/GenerateCodeHelper.cs
    28      0       src/DocoptNet/GeneratedSourceModule.cs
    1       1       src/DocoptNet/Pattern.cs
    279     106     src/DocoptNet/PatternMatcher.cs
    2       0       src/DocoptNet/ValueObject.cs

`PatternMatcher.cs` was refactored extensively (by consolidating duplication) so that the generated code could _efficiently_ share as much of the pattern matching logic as possible with the runtime/library version.

Apart from the technical changes, reviewers should focus on the overall design of the generated code in terms of:

1. Simplicity
2. Experience
3. Above all, hindering any customisation or extensibility in the future

**Comment by voieducode** [comment]
Super helpful notes, thank you @atifaziz

**Comment by atifaziz** [comment]
In my own review, I noticed that the `exit` argument of `Apply` was never used in the generated code and the implementation always acted as if `exit: true` was the supplied argument. With, fa0c013f9b8ba2669a257866e3df91c9313d04c5 (+ 783d8132d438fc5b0daf014a08375aae8270031c), I've removed the `exit` parameter entirely. Eventually, I think all the parameters will go away except `args`, but those tweaks can come in post merge of this PR.

**Review by voieducode — APPROVED** [review]
After our offline conversation about the PR, I have no additional questions or comments - thanks!

File: eg/SourceGenerator/ArgumentsExample/Program.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+}
+catch (DocoptExitException e)
+{
+    Console.WriteLine(e.Message);
+    return e.ErrorCode;
+}
+catch (DocoptInputErrorException e)
+{
+    Console.Error.WriteLine(e);
+    return 0xbd;
```
[the remark this answers is not in the document]
Comment (atifaziz) [reply eg/SourceGenerator/ArgumentsExample/Program.cs:19]: Not at all. It's a very arbitrary exit code and something I typically use. `0xbd` basically is `0xbad` but with the `a` removed to fall in the 0-255 range.

File: eg/SourceGenerator/ArgumentsExample/obj/Generated/DocoptNet.CodeGeneration/DocoptNet.CodeGeneration.SourceGenerator/ProgramArguments.cs [hunk 2]
```diff
@@ -0,0 +1,290 @@
+#nullable enable annotations
```
[the remark this answers is not in the document]
Comment (atifaziz) [reply eg/SourceGenerator/ArgumentsExample/obj/Generated/DocoptNet.CodeGeneration/DocoptNet.CodeGeneration.SourceGenerator/ProgramArguments.cs:1]: I think #76 kind of addressed that already.

File: eg/SourceGenerator/Git/Git.docopt.txt [hunk 3]
```diff
@@ -0,0 +1,21 @@
+usage: git [--version] [--exec-path=<path>] [--html-path]
```
[the remark this answers is not in the document]
Comment (atifaziz) [reply eg/SourceGenerator/Git/Git.docopt.txt:1]: Did you see [my reply there](https://github.com/docopt/docopt.net/issues/114#issuecomment-912501001)?

**Comment by voieducode** [comment]
Congratulations for the merge!!! 🎉🎉🎉

**Comment by atifaziz** [comment]
> Congratulations for the merge!!! 🎉🎉🎉

Thanks! Slowly, but surely, made it to the finishing line! 🏁 Appreciate all your trust, patience and support during the months it took to work through this PR! 😄

**Pull request merged**